### ISSUE TITLE
Limit number of cores for HMSC

### ISSUE THREAD

**Issue by kdanielmorais** [hmsc-r/HMSC#23]
Dear colleagues,

When generating the models, how can I limit the number of cores used by HMSC. I don't seem to find a parameter for that. When I put it to run, it seems to use all the cores available in our server.

Here is the command that seems to be using all the cores:
sampleMcmc(m, samples = samples, thin=thin, adaptNf=rep(ceiling(0.4*samples*thin),1), transient = ceiling(0.5*samples*thin), nChains = nChains, nParallel = nChains)

Many thanks in advance,
Daniel

**Comment by jarioksa** [comment]
You specified it to use `nParallel = nChains` cores. Does it use more cores than you requested?

**Comment by kdanielmorais** [comment]
Yes. nChains is equal to 4 in this case and the script is using 80 cores (all available in the machine). We are running it in a linux server, but I could not find any other parameter to limit the number of cores.

**Comment by gtikhonov** [comment]
I think the problem you are running here is due to some quite specific build of the linear algebra libraries that your R uses. We had some similar reports before. To validate that this is the case, you can try 2 things:
1) Fit your model with `nParallel = 1` and check how many cores the code uses.
2) Just do some large matrix operation in a loop and check how many cores does it exploit. If the number is far beyond one, the issue is confirmed. E.g. something like
```
R = 100
N = 2001
A = matrix(rnorm(N^2),N,N)
B = crossprod(A)
for(r in 1:R){
  W = B + diag(1+runif(N))
  iW = solve(W)
  print(sprintf("%d - %f", r, sum(iW)))
}
```

**Comment by kdanielmorais** [comment]
Dear Gleb,

Many thanks for the help. It seems you are right. In the first check, using nParallel = 1, the computer was using 31 cores. In the second test, with the loop, all the cores were used.

Would you have any suggestion to have more control over the number of cores?

**Comment by gtikhonov** [comment]
Well, I would suggest to start from having a look here
https://cran.r-project.org/doc/manuals/R-admin.html#Linear-algebra
A small disclaimer - that time we actually avoided solving this issue... But I fully agree that this is an important issue that should be properly addressed. 
Maybe @jarioksa has a better insight than me, any thoughts?

**Comment by MelindadeJonge** [comment]
Hi Daniel, 

There is also a package on CRAN specifically made to control the number of threads used by BLAS routines called 'RhpcBLASctl' (https://cran.r-project.org/web/packages/RhpcBLASctl/index.html). I just tested this on our linux server using Gleb's example and it seems to work nicely. 
Alternatively, you may be able to limit the number of cores available to your job through your scheduling system which also worked for me. 

Good luck!

**Comment by kdanielmorais** [comment]
Perfect, Melinda.
I will use this package. I made a test with Gleb's large matrix operation loop and it worked well. 

Many thanks, Jari, Gleb and Melinda for the very fast response.

Best regards,

**Comment by kdanielmorais** [comment]
Sorry for reopening it here folks. Although it worked with the loop aforementioned, when we tried with the sampleMcmc() it still uses all the cores. What we are doing so far is to run nParallels = 2 and this keeps the load in a limit of ~ 60 cores. But thanks for the support.

**Comment by MelindadeJonge** [comment]
Dear Daniel, 

You closed this issue again right after reopening. Did you find a way to fix your issue?

Best regards, 
Melinda

**Comment by gtikhonov** [comment]
But did the above mentioned package fix this issue when using a single core? If yes, then the problem originates from the way how the multiple threads are implemented - these are effectively new separate processes, which load a set of their own packages.
https://github.com/hmsc-r/HMSC/blob/7e55207e314ce7c10215fee64b799581c24850ac/R/sampleMcmc.R#L287
Then I would guess that it is required to repeat the same controlling procedure for each of these processes. Could you post your adjusted code for that toy example that I used above, which code solved your issue? I will try to adapt it to be included to the HMSC package, so that the user would be capable to control the amount of cores used with BLAS by passing it as an argument to the `sampleMcmc(...)` function.

**Comment by jarioksa** [comment]
Sparse matrix operations replace BLAS with other BLAS-like routines. Any idea if these implement parallel processing?

**Comment by kdanielmorais** [comment]
Hi everyone,

Thanks for reopening it. This issue is still not solved.

To your questions:

1. The package 'RhpcBLASctl' doesn't limit the number of clusters when running sampleMcmc(... nParallels = 1).
2. The adjusted code used with the little for loop example is:

```{r}
library(RhpcBLASctl) 
omp_set_num_threads(10)
blas_set_num_threads(10)
R = 100
N = 2001
A = matrix(rnorm(N^2),N,N)
B = crossprod(A)
for(r in 1:R) {
  W = B + diag(1+runif(N))
  iW = solve(W)
  print(sprintf("%d - %f", r, sum(iW)))
}
```
But this code, although controlled the number of cores in this example, didn't affect the sampleMcmc() function. We set omp_set_num_threads(5)
blas_set_num_threads(5) and even though when running sampleMcmc(... nParallels = 1) the machine sitll used 31 cores.

**Comment by gtikhonov** [comment]
Hm...
> We set omp_set_num_threads(5)
blas_set_num_threads(5) and even though when running sampleMcmc(... nParallels = 1) the machine sitll used 31 cores.

Such misbehaviour really confuses me - this means that the "for loop example" fails to properly mimic HMSC's numeric routines. I will try to replicate this issue using the resources that are available to me now. You may greatly facilitate me if you could uncover the model structure you are trying to fit. Effectively, I need to see the arguments of the calls to `Hmsc(...)` and `HmscRandomLevel(...)` functions and preferably also the dimensionality of the data objects that you pass to these calls (I do not ask for the objects themselves).

**Comment by kdanielmorais** [comment]
Gleb,

Here are the details you asked:
There are only sample level random effects. “Sample” has unique 
identifiers per each observation.
```
HmscRandomLevel(units = studyDesign$Sample)

m = Hmsc(Y=Y,
          XData = XDataList,  XFormula = XFormula,
          TrData = TrData, TrFormula = TrFormula,
          distr={"probit"} ,
          studyDesign=studyDesign, ranLevels={list(Sample=rL.Sample)})
for (i in (ns+1):(2*ns)){
             m$distr[i,1:2] = c(1,1)
          }
```
Y is a data.frame with species of bacteria (15) and fungi (21). 
Presence-absence data and abundance data (conditional on presence) are 
juxtaposed. So that:
```
dim(Y)
[1] 118 72
```
XDataList is a list of data.frames of length 72. One explanatory 
variable changes depending on whether the response species is bacteria 
or fungi.

TrData includes a factor with four levels indicating if the species is 
bacteria or fungi, and p-a or abundance data.

The last loop changes the default “probit” link to lognormal in the 
columns with abundance data.

This script is being used by a colleague in our server, so I will copy him here. If we need more details he can jump in the discussion. @odrilarra

Thanks

**Comment by jarioksa** [comment]
The same thing seems to happen in our Ecology Linux here in Helsinki as well. However,  with us this is not an **Hmsc** issue, but an issue of **LAPACK** configuration in **R**. It is certain that **Hmsc** does not launch more than `nChains` parallel processes: they are chains, and chain is a sequence and  we cannot jump in the middle of a chain. However, we use heavily libraries for matrix algebra, and these can be implemented to use parallel processing. In our case we see:
```
> sessionInfo()
...
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
```
Here the **LAPACK** library is multithreaded (the last `p` in `libopenblasp` stands for `pthreads`). Please check your **BLAS**/**LAPACK**. For instance, Cholesky factorization can be implemented with parallel processing. Cholesky takes most of processing time in typical applications, and parallel processing will help a lot. The **R** can be configured to use several alternative **BLAS**/**LAPACK** implementations, and changing to another one can make a great difference – both in the use of CPUs and in processing time. Alternatively, you should consult your sysadmin for configuring your **LAPACK**. This is not something we can do from within **Hmsc**.

**Comment by CescGV** [comment]
I had this problem for a long time, and I found a very easy and hopefully okayish solution. I limit the number of threads to 1, using the package RhpcBLASctl and options(mc.cores= 1), and then I independently run X tmux sessions , each of them with the same Hmsc model fitted using sample Mcmc (where X is the previous nChains). Then I independently save each fitted model, and gather them afterwards. 

In summary, I "manually parallelize" the process. It's a bit unpractical, but it solves the issue.
